# **tyPPing** and other methods for detecting phage-plasmids

**For further information, see: "Efficient detection and typing of phage-plasmids"** [ADD LINK]

Here you can find the documentation, scripts, and data files necessary for the **identification of phage-plasmids (P-Ps)** (folders named after each approach). 
Data and scripts required to reproduce analysis and figures of our study are in 'Publication_related_data'.

For detailed information see **`DESCRIPTION_XXX.md`**

## Repository structure:

### *tyPPing*

tyPPing is a user-friendly, fast and accurate method, designed for the sensitive and efficient detection of distinct P-P types. Currently it finds P-Ps of the type AB_1, P1_1, P1_2, N15, SSU5, pMT1, pCAV, pSLy3, pKpn, and cp32.
It uses protein profiles to search sequences for patterns (frequency and compositional sets) of conserved P-P proteins. If a match also fits the typical size range, it is predicted as a P-P with a distinct confidence.

`DESCRIPTION_tyPPing.md` describes how to use tyPPing for P-P detection.

### *MM-GRC*

MM-GRC (**m**ulti-**m**odel **g**ene **r**epertoire **c**lustering) is our first method to classify P-Ps, that we developed in PMID: 33590101. It is an integrated approach combining functional annotation (with phage- and plasmid-specific HMM profiles), machine learning (Random Forest, RF) models, and which was complemented with a literature review. It relies on the gene repertoire relatedness to type P-Ps. It detects various P-P types including diverse communities and unrelated putative P-Ps (singletons).

`DESCRIPTION_MM-GRC.md` describes how to detect and type P-Ps using MM-GRC.

### *geNomad_vConTACT2*

This approach combines the use of geNomad and vConTACT v2 for the identification and typing of P-Ps.

geNomad is a computational tool for classifying nucleotide sequences as phages, integrated prophages, or plasmids [REF]. In this study, it is used to analyze plasmid sequences and identify candidates with phage-like characteristics, indicating potential P-Ps.

vConTACT v2 is a viral classification tool that clusters viral genomes based on shared gene content [REF]. Here, it is used to further group the P-P candidates identified by geNomad. Plasmid genomes predicted to be phage-like are compared with a reference set of known P-Ps to determine their type based on similarity.

`DESCRIPTION_geNomad_vConTACT2.md` provides instructions for using geNomad and vConTACT v2 to detect P-Ps.

### *Publication_related_data*

This folder contains the scripts, data, and supplementary materials needed to reproduce the figures and analyses presented in the publication [REF].

`DESCRIPTION_publication_data.md` provides more detailed information.

## Files stored in Zenodo repository

Due to storage limits, several essential files are available for download from a Zenodo repository: **[Link to the repository]**

-   `tyPPing_signature_profiles.hmm` – HMM profiles specific to P-P types, used for protein-to-profile searches by tyPPing.

-   `phage.hmm` – phage-specific HMM profiles used in MM-GRC pipeline.

-   `models/` – a folder containing RF classification models trained to detect P-Ps in plasmid datasets (used by MM-GRC).

-   `g2g_plot_tables/` and `tyPPing_criteria_tables/` contain the input tables required to run the scripts `all_g_to_g_plots_filtered.R` and `figures_methods.R` from `Publication_related_data/` folder.

-   `draft_genomes_analysis/` – a folder containing 12 selected draft genomes from carbapenem-resistant *Enterobacteriales* (CRE) species (including short-read assemblies, long-read assemblies, and complete P-P genomes in corresponding subfolders) and prediction tables generated by `tyPPing_for_draft_genomes.R.`

## References

-   [MM-GRC](https://pmc.ncbi.nlm.nih.gov/articles/PMC7969092/)
-   [HMMER](https://github.com/EddyRivasLab/hmmer)
-   [MMseqs2](https://github.com/soedinglab/MMseqs2)
-   [geNomad](https://portal.nersc.gov/genomad/index.html)
-   [vConTACTv2](https://bitbucket.org/MAVERICLab/vcontact2/src/master/)
-   [PHROGs](https://phrogs.lmge.uca.fr/)

## Citing tyPPing

If you use tyPPing, please cite the corresponding paper:
