# Introduction

**tyPPing** is an R-based pipeline designed for fast and accurate detection and typing of phage-plasmids (P-Ps). It uses a curated set of protein profiles trained on well-known P-P types to find conserved signature proteins and classify elements accordingly. The pipeline evaluates both the presence and unique patterns of these proteins, assigning a P-P type along with a confidence score. This allows users to quickly identify new P-Ps, making tyPPing a practical tool for large-scale P-P screening and systematic classification.

Currently, tyPPing can detect the following P‑P types:

**AB_1, P1_1, P1_2, N15, SSU5_pHCM2, pMT1, pCAV, pSLy3, pKpn, and cp32.**

The pipeline works with both **complete** and **draft genomes**, using two dedicated scripts:

-   **`tyPPing.R`** – for complete genomes.
-   **`tyPPing_for_draft_genomes.R`** – for draft genomes (contigs or MAGs).

## Pipeline overview

**Figure guide**

<img width="1241" height="437" alt="pipeline" src="https://github.com/user-attachments/assets/557b1031-792a-4985-970d-6a1badb19e07" />

**Step 1. Protein search**

Use **HMMER** (`hmmsearch`) to compare your multi-protein FASTA file against the provided P‑P signature profiles.

**Step 2. Run tyPPing (R script)**

After that, tyPPing script processes the `--domtblout` output from the previous step through two analytical branches:

-   **MinProteins** checks how many strongly conserved P-P proteins are found in each element and keeps only those that have at least a minimum threshold number.
-   **Composition** keeps elements only if they match one of the unique sets of conserved proteins defined for a given P-P type.

Detection results obtained by these branches are used together with genome size to assign each genome a predicted P-P type and a confidence level (**High**, **Medium**, or **Low**). The results are summarized in `Final_prediction_table.tsv`.

**Folder structure**

The **`tyPPing/`** folder contains the main scripts, input tables, and example data:

```{bash eval=FALSE, include=FALSE}
tyPPing/
├── tyPPing.R                    # Main script for complete genomes
├── tyPPing_for_draft_genomes.R  # Script for draft genomes (e.g., MAGs)
├── tyPPing_input_data/
│   ├── Cutoffs_information_table.tsv      # tyPPing score thresholds 
│   ├── Compositions_information_table.tsv # Profile composition sets per P-P type
│   └── Profile_information_table.tsv      # Annotation information (used by tyPPing)
├── small_example_test_data/     # Includes example input/output for 20 genomes
```

# Usage

## Setup & installation

-   **tyPPing** is an R-based tool and does not require installation. To use it, simply **clone** or download the repository.

-   **Requirements:**

    -   **HMMER** must be installed and accessible via the command line (used for hmmsearch) [REF].

    -   **R** installed (version ≥ 4.0) [link].

    -   "rhmmer" R package is required (not available on CRAN) [link]. You can install it using:

        ```{r eval=FALSE, include=FALSE}
        devtools::install_github('arendsee/rhmmer')
        ```

-   Download the P-P type specific HMM profiles file **`tyPPing_signarture_profiles.hmm`** from [Zenodo link].

## Input data requirements

**1. For HMM search (Step 1):**

-   A multi-protein FASTA file sutable for the hmmsearch (e.g., `proteins.fasta`).

-   P-P type-specific HMM profiles file `tyPPing_signarture_profiles.hmm`.

**2. For tyPPing scripts (Step 2):**

To run `tyPPing.R` (for complete genomes) or `tyPPing_for_draft_genomes.R` (for draft genomes), you will need:

-   HMMER output table from Step 1.

-   tyPPing input tables provided in the `tyPPing_input_data/` folder.

-   A protein-to-genome mapping file (generated by user, see example below).

-   A genome size file (generated by user, see example below).

## **Running tyPPing**

#### 1. Search for P-P specific proteins in target genomes

Run `hmmsearch` (from HMMER) to scan the protein sequences of target genomes (e.g., `proteins.fasta`) against the P‑P type‑specific profile HMMs (`tyPPing_signarture_profiles.hmm`). For detailed HMMER documentation, see [REF].

Example command:

```{bash eval=FALSE, include=FALSE}
hmmsearch -o tmp.all.out --domtblout hmm_search_for_tyPPing.tbl.out tyPPing_signarture_profiles.hmm proteins.fasta
```

#### 2. Prepare the input files

tyPPing requires two simple input tables in TSV or CSV format:

-   **Protein‑to‑genome** mapping file (e.g., `protein_to_genome.tsv`). This table links each protein ID to its corresponding genome. It must have 2 columns named:

    -   `protein_id` – protein identifier used in the FASTA and HMMER output

    -   `genome_id` – genome identifier for each protein.

    Example table:

    | protein_id | genome_id |
    |------------|-----------|
    | NP_052607  | NC_002128 |
    | NP_052608  | NC_002128 |
    | NP_052609  | NC_002128 |
    | NP_052610  | NC_002128 |
    | ...        | ...       |

-   **Genome size** file (e.g., `genome_size.tsv`). This table contains the genome sizes and must have 2 columns named:

    -   `genome_id` – genome identifier

    -   `size` – genome size in base pairs (bp).

    Example table:

    | genome_id | size  |
    |-----------|-------|
    | NC_002128 | 92721 |
    | NC_005856 | 94800 |
    | NC_006509 | 47890 |
    | NC_015465 | 33004 |
    | ...       | ...   |

Once these two files and the `--domtblout` HMMER output are ready, you can proceed to run the tyPPing script.

#### 3. Run the tyPPing script

You can run **`tyPPing.R`** (for complete genomes) or **`tyPPing_for_draft_genomes.R`** (for draft genomes) in two ways:

#### **A. Interactive mode (in RStudio):**

-   Update all file paths in the script to match your local directory structure (see the script for guidance).

-   Run the script.

**B. Command‑line mode:**

```{bash}
    Rscript tyPPing.R\    # or tyPPing_for_draft_genomes.R
    --map path/to/protein_to_genome.tsv \
    --sizes path/to/genome_size.tsv \
    --hmm_domtbl path/to/all_pers_hmm_plasmids_0523_out.tbl.out \
    --outdir path/to/output_dir/ \
    --compositions path/to/Compositions_information_table.tsv \
    --profiles path/to/Profile_information_table.tsv \
    --cutoffs path/to/Cutoffs_information_table.tsv
```

**Command‑line required arguments:**

-   `-m, --map` – path to *protein‑to‑genome* mapping table.

-   `-s, --sizes` – path to *genome size* table.

-   `-i, --hmm_domtbl` – path to *HMMER output file* (`*.tbl.out`).

-   `-o, --outdir` – path to the *output* directory (must be created before running tyPPing).

-   `--compositions` – profile composition sets for each P‑P type (by default in: `tyPPing_input_data/Compositions_information_table.tsv`).

-   `--profiles` – profile metadata, functional annotations, and mapping between HMM IDs and P-P types (by default in: `tyPPing_input_data/Profile_information_table.tsv`).

-   `--cutoffs` – thresholds used by the MinProteins and Composition branches per P‑P type (by default in: `tyPPing_input_data/Cutoffs_information_table.tsv`).

## Standard output summary tables

tyPPing generates **two output tables**:

**1. `Final_prediction_table.tsv`**

This table contains the **final predictions** for elements detected by at least one branch (**MinProteins** and/or **Composition**) and includes confidence levels for each assignment.

<img width="1146" height="173" alt="prediction" src="https://github.com/user-attachments/assets/d38527cf-3786-4dac-adfc-30d7e657388e" />

**Column descriptions:**

-   **Genome ID** – unique identifier of the analyzed genome.
-   **Genome size (bp)** – total size of the genome, in base pairs.
-   **P-P type** – the predicted phage-plasmid (P-P) type assigned to the genome.
-   **Confidence level** – prediction confidence (**High**, **Medium**, or **Low**) based on the number of branches supporting the result and genome size.
-   **Predicted by** – branch or branches that led to the prediction.
-   **MinProteins cutoff** – minimum number of required protein hits for a confident match to the P-P type (for MinProteins branch).
-   **MinProteins hits** – number of signature proteins detected by the **MinProteins** branch.
-   **Composition** – composition pattern ID (set of signature profiles) matched for the P-P type (for Composition branch).
-   **Composition size** – total number of signature profiles in the defined composition set.
-   **Composition hits** – number of composition-specific profiles found in the genome by the **Composition** branch (considering tolerance to gaps).

**2. `All_hmm_hits_table.tsv`**

This table lists **all HMM hits** (not just confident predictions). It is useful for inspection of intermediate data or investigating borderline cases.

<img width="1160" height="125" alt="all_hits" src="https://github.com/user-attachments/assets/c26bc977-cc86-4ada-b94e-b58e73815397" />

-   **Genome ID** – unique identifier of the analyzed genome.
-   **P-P type** – phage-plasmid (P-P) type for which the hits are analyzed (can be several entries per genome).
-   **MinProteins cutoff** – minimum number of required protein hits for a confident match to the P-P type (for MinProteins branch).
-   **MinProteins hits** – number of signature proteins detected by the **MinProteins** branch.
-   **Signature to all genes ratio** – ratio of detected signature proteins to the total number of proteins in the genome.
-   **P-P score sum** – sum of *P-P scores* for all matched signature profiles.
-   **P-P type score sum** – sum of *P-P type scores* for all matched signature profiles.
-   **Composition** – composition pattern ID (set of signature profiles) matched for the P-P type (for Composition branch).
-   **Composition size** – total number of signature profiles in the defined composition set.
-   **Tolerance to gaps** – maximum number of missing profiles allowed for a valid composition match.
-   **Composition hits** – number of composition-specific profiles found in the genome by the **Composition** branch.
-   **Genome size (bp)** – total size of the genome, in base pairs.
-   **Number of proteins** – total number of protein-coding genes in the genome.

## Modification of tyPPing to detect P-Ps in draft genomes and MAGs

To make tyPPing suitable for incomplete or fragmented genomes, we updated its scoring system to work at the genome level rather than per sequence. This means conserved P-P proteins are counted across all contigs in a genome. With this change, tyPPing can detect P-Ps that are split across multiple contigs with a high density of signature genes, as long as this set of contigs meets the MinProteins and/or Composition thresholds.

**Use the dedicated script `tyPPing_for_draft_genomes.R.`**

**Input tables** provided by the user are different. For this version, both contig and genome IDs are required:

-   *Protein-to-genome* mapping file example:

    | protein_id       | contig_id      | genome_id |
    |------------------|----------------|-----------|
    | 170D8_contig_1_1 | 170D8_contig_1 | 170D8     |
    | 170D8_contig_1_2 | 170D8_contig_1 | 170D8     |
    | 170D8_contig_1_3 | 170D8_contig_1 | 170D8     |

-   *Contig sizes* file example:

    | contig_id       | size    | genome_id |
    |-----------------|---------|-----------|
    | 170D8_contig_1  | 5091073 | 170D8     |
    | 170D8_contig_13 | 198075  | 170D8     |
    | 170D8_contig_14 | 34813   | 170D8     |

**Output tables** `Final_prediction_table.tsv` and `All_hmm_hits_table.tsv` include several additional output columns:

-   **MinProteins hits list** – the list of counts showing how many signature proteins for the predicted P-P type were found in each contig by MinProteins branch, separated by ";".

-   **MinProteins contigs list** – the list of contigs where MinProteins hits were found, corresponding in order to the counts in MinProteins hits list, separated by ";".

-   **Composition contigs list** – the list of contigs where composition-specific profiles (from Composition branch) were detected, separated by ";".

**There are several limitations of this version:**

-   This approach does not identify multiple P-Ps of the same type within a single genome. However, since we expect P-Ps of the same type to be likely incompatible (like plasmids), these events should be rare.

-   It can still detect P-Ps of different types in the same genome, thanks to the type-specific signature profiles used by tyPPing.

-   For MAGs, accurate binning is crucial. If contigs are mis-binned, results may be incorrect or chimeric P-P predictions.

## **tyPPing's performance notes**

-   **Detection accuracy:**
    -   \>99% sensitivity (especially for cp32 P-Ps) and \>99% precision compared to MM-GRC.
-   **Running time:**
    -   \~7 minutes to process \>38,000 plasmids (05/23 dataset) with `tyPPing.R` and additional \~1h44m for the HMM protein-to-profile comparison using HHMER.

## **Interpretation and recommendations**

-   You can test the workflow using the small dataset in `small_example_test_data/`. It includes 20 genomes and all required inputs and outputs.

-   **High confidence** predictions are detected by **both branches** (*MinProteins* and *Composition*) and fit the expected **size** range. These are the most reliable predictions but may miss \~20% of P-Ps if only this category is considered.

-   **Medium confidence** predictions are detected by one or both branches and fit the size range with 10% tolerance. These often represent more divergent P-Ps (such as many cp32-like). Rare false positives (three elements in 05/23 dataset) were detected with Medium confidence.

-   **Low confidence** is assigned to elements with genome sizes significantly shorter or longer than expected. Short genomes may indicate degrading P-Ps, while unusually long ones may result from recombination. Although tyPPing reports these elements, they should not be considered reliable predictions without further manual inspection.

# Citing tyPPing

-   If you use tyPPing, please cite the corresponding paper:
